Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py

```diff
@@ -0,0 +1,714 @@
+                                                                                    icol, use_nearest)
+                        if not np.any(np.logical_and(opt_peak_mask, det_edges)) or integrate_on_edge:
+                            # combine masks as optimal TOF window can truncate peak slightly
+                            opt_peak_mask = np.logical_or(peak_mask, opt_peak_mask)
+                            is_valid_mask, new_status = is_peak_mask_valid(opt_peak_mask, peak_label, npk_min,
+                                                                           density_min, nrow_max, ncol_max,
+                                                                           min_npixels_per_vacancy, max_nvacancies)
+                            if is_valid_mask:
+                                status = new_status
+                                # refocus data and re-optimise TOF limits
+                                peak_mask = opt_peak_mask
+                                non_bg_mask = np.logical_and(non_bg_mask, opt_non_bg_mask)
+                                ypk, epk_sq = focus_data_in_detector_mask(signal, error, peak_mask, non_bg_mask, ixpk)
+                                ixlo_opt, ixhi_opt = find_peak_limits(ypk, epk_sq, min(ixlo, ixlo_opt),
+                                                                      max(ixhi, ixhi_opt))
+                    if update_peak_pos:
+                        # find det
+                        idet = np.argmax(signal[:, :, ixlo_opt:ixhi_opt].sum(axis=2)[peak_mask])
+                        det = dets[peak_mask][idet]
+                        irow, icol = np.where(dets == det)
+                        # update tof
+                        imax = np.argmax(ypk[ixlo_opt:ixhi_opt]) + ixlo_opt
+                        # replace last added peak
+                        irows_delete.append(pk_ws_int.getNumberPeaks() - 1)
+                        self.child_AddPeak(PeaksWorkspace=pk_ws_int, RunWorkspace=ws, TOF=xpk[imax],
+                                           DetectorID=int(det))
+                        pk = pk_ws_int.getPeak(pk_ws_int.getNumberPeaks() - 1)
+                    # normalise by bin width before final integration
+                    dx = np.diff(xpk)
+                    ypk[1:] = ypk[1:] * dx
+                    ypk[0] = ypk[0] * dx[0]  # assume first has same dx as adjacent bin
+                    epk_sq[1:] = epk_sq[1:] * (dx ** 2)
+                    epk_sq[0] = epk_sq[0] * (dx[0] ** 2)
+                    # calc intens and sig
+                    th = pk.getScattering() / 2
+                    wl = pk.getWavelength()
+                    L = (np.sin(th) ** 2) / (wl ** 4)  # at updated peak pos if applicable
+                    intens = L * np.sum(ypk[ixlo_opt:ixhi_opt])
+                    sig = L * np.sqrt(np.sum(epk_sq[ixlo_opt:ixhi_opt]))
+                    integrated = True
```

Comment:
The exec function is quite large at the moment. Perhaps you could pull out L589-637 into a function? If there's lots of variables need using after this point then not a "must do" but worth checking if this is possible